src/PixiEditor.ChangeableDocument/Changeables/Graph/Nodes/CombineSeparate/SeparateColorNode.cs

```diff
@@ -19,39 +22,61 @@ public class SeparateColorNode : Node
     
     public FuncOutputProperty<Float1> A { get; }
     
-    
-    private FuncContext lastContext;
-    private Half4 lastColor;
-
     public SeparateColorNode()
     {
-        Color = CreateFuncInput<Half4>(nameof(Color), "COLOR", new Color());
         R = CreateFuncOutput<Float1>(nameof(R), "R", ctx => GetColor(ctx).R);
         G = CreateFuncOutput<Float1>(nameof(G), "G", ctx => GetColor(ctx).G);
         B = CreateFuncOutput<Float1>(nameof(B), "B", ctx => GetColor(ctx).B);
         A = CreateFuncOutput<Float1>(nameof(A), "A", ctx => GetColor(ctx).A);
+        Mode = CreateInput("Mode", "MODE", CombineSeparateColorMode.RGB);
+        Color = CreateFuncInput<Half4>(nameof(Color), "COLOR", new Color());
     }
 
     protected override Texture? OnExecute(RenderingContext context)
     {
         return null;
     }
     
-    private Half4 GetColor(FuncContext ctx)
+    private Half4 GetColor(FuncContext ctx) =>
+        Mode.Value switch
+        {
+            CombineSeparateColorMode.RGB => GetRgba(ctx),
+            CombineSeparateColorMode.HSV => GetHsva(ctx),
+            CombineSeparateColorMode.HSL => GetHsla(ctx)
+        };
+
+    private Half4 GetRgba(FuncContext ctx) => ctx.Builder.GetOrNewAttachedHalf4(this.GetHashCode(), Color.GetHashCode(), () => Color.Value(ctx));
+
+    private Half4 GetHsva(FuncContext ctx)
     {
-        Half4 target = null;
-        if (lastContext == ctx)
+        if (!ctx.HasContext && ctx.GetValue(Color) is Half4 constantColor)
         {
-            target = lastColor;
+            var variable = new Half4(string.Empty);
+            constantColor.ConstantValue.ToHsv(out float h, out float s, out float l);
+            variable.ConstantValue = new Color((byte)(h * 255), (byte)(s * 255), (byte)(l * 255), constantColor.ConstantValue.A);
```

Comment:
So the variable is Hsva color value, but constant values are multiplied like it's rgb?

Reply:
if it wasn't multiplied you would always get a output of either O or 1 per component